### PR TITLE
#47 spring data repository support

### PR DESCRIPTION
**Pull request by gessnerfl**
This is my first implementation of the spring data repository support. I implemented everything in a new gradle project so that users without spring-data in use do not get spring data into the classpath by exception.
The aspect is implemented with AspectJ. I first tried to implement it with a BeanPostProcessor but this did not work as the target repository instance is not available at that time but only its own factory. Regarding https://jira.spring.io/browse/DATAMONGO-817 BeanPostProcessors are not supported as the repositories itself rely on this feature.

### CODE REVIEW

File: javers-core/src/main/java/org/javers/common/collections/Optional.java [hunk 1]
```diff
@@ -54,6 +54,13 @@ public T get(){
         return reference;
     }
 
+    public T getOrElse(T nullValue){
```
Comment (pawelszymczyk) [inline javers-core/src/main/java/org/javers/common/collections/Optional.java]: this method is never used

File: javers-spring-data/src/main/java/org/javers/spring/data/aspect/JaversSpringDataRepositoryAspect.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import org.springframework.data.repository.core.support.DefaultRepositoryMetadata;
+
+import java.util.Collection;
+import java.util.Collections;
+
+@Aspect
+public class JaversSpringDataRepositoryAspect {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(JaversSpringDataRepositoryAspect.class);
+    private final Javers javers;
```
Comment (pawelszymczyk) [inline javers-spring-data/src/main/java/org/javers/spring/data/aspect/JaversSpringDataRepositoryAspect.java]: javers and authorProvider could be a local variable:

```
 private final AuditChangeHandler saveHandler;
    private final AuditChangeHandler deleteHandler;

    public JaversSpringDataRepositoryAspect(Javers javers) {
        this(javers, new DefaultAuthorProvider());
    }

    public JaversSpringDataRepositoryAspect(Javers javers, AuthorProvider authorProvider) {
        this(new OnSaveAuditChangeHandler(javers, authorProvider), new OnDeleteAuditChangeHandler(javers, authorProvider));
    }

    protected JaversSpringDataRepositoryAspect(AuditChangeHandler saveHandler, AuditChangeHandler deleteHandler) {
        this.saveHandler = saveHandler;
        this.deleteHandler = deleteHandler;
    }
```

File: javers-spring-data/src/main/java/org/javers/spring/data/aspect/JaversSpringDataRepositoryAspect.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private Optional<Class> getRepositoryInterface(ProceedingJoinPoint pjp) {
+        for (Class i : pjp.getTarget().getClass().getInterfaces()) {
+            if (i.isAnnotationPresent(JaversSpringDataAuditable.class) && CrudRepository.class.isAssignableFrom(i)) {
+                return Optional.of(i);
+            }
+        }
+        return Optional.empty();
+    }
+
+    private Iterable<Object> getObjectList(ProceedingJoinPoint pjp) {
```
Comment (pawelszymczyk) [inline javers-spring-data/src/main/java/org/javers/spring/data/aspect/JaversSpringDataRepositoryAspect.java]: maybe sth more descriptive, eg ``` getJoinPointArguments ```

File: javers-spring-data/src/main/java/org/javers/spring/data/handler/AuditChangeHandler.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.javers.spring.data.handler;
+
+import org.springframework.data.repository.core.RepositoryMetadata;
+
+/**
+ * Created by gessnerfl on 22.02.15.
+ */
+public interface AuditChangeHandler {
+    void onAfterRepositoryCall(RepositoryMetadata repositoryMetadata, Object changedObject);
```
Comment (pawelszymczyk) [inline javers-spring-data/src/main/java/org/javers/spring/data/handler/AuditChangeHandler.java]: maybe just ```handle(...)``z with some descriptive javadoc?

**Comment by pawelszymczyk** [comment]
It looks very well, I don't have more comments.

**Comment by gessnerfl** [comment]
Thanks for your feedback. Clean up the code according your review.

**Comment by pawelszymczyk** [comment]
Excellent, thanks for great contribution!

**Comment by bartoszwalacik** [comment]
to be released tomorrow in v1.0.7

**Pull request merged**